Introduce `bin/rails clear` command for clearing logs

p8 opened the pull request:
Currently we use both Thor and Rake for `bin/rails` commands.
We eventually want to get all the built-in tasks promoted to Thor Commands:
https://github.com/rails/rails/pull/47610#issuecomment-1461520271
This migrates the `log:clear` task to Thor.

Currently logs and the tmp directory can be cleared with a single invocation:

```bash
bin/rails log:clear tmp:clear
```

With Thor we can't run multiple commands with a single invocation, as
additional commands will be seen as arguments to the first command.

This introduces a single `clear` command that will combine the existing
`log:clear` and `tmp:clear` commands. For now it's limited to clearing logs:

```bash
bin/rails clear --logs
```

The `--logs` option also supports passing `all` or a list of environments:

```bash
bin/rails log:clear --logs all # similar to `bin/rails log:clear LOGS=all`
bin/rails log:clear --logs development test # similar to `bin/rails log:clear LOGS=development,test`
```

Thor doesn't support Rake's FileList, so for truncating all log files we
use a glob pattern and filter out directories.


### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

rafaelfranca commented on railties/lib/rails/tasks.rb:
```diff
@@ -5,7 +5,6 @@
 # Load Rails Rakefile extensions
 %w(
   framework
-  log
```
Why we are not removing the `.rake` file that define this command?

p8 replied:
Thanks, I missed that.
I've removed the file.

skipkayhil commented:
Same concern as https://github.com/rails/rails/pull/47619#discussion_r1131745563 I think this could allow for modifying files outside of the application folder now

rafaelfranca commented:
That is fine. Rake would as well, it just don't right now because the Rakefile can't be loaded.

p8 commented:
This now also works outside the app directory so we can do:
```bash
BUNDLE_GEMFILE=demo/Gemfile demo/bin/rails log:clear
```

I've also fixed the case for passing "LOGS=all".
I wanted to add a test but it's currently pretty difficult to pass ENV vars to the `rails` test helper.

jonathanhefner reviewed (commented):
> I've also fixed the case for passing "LOGS=all". I wanted to add a test but it's currently pretty difficult to pass ENV vars to the `rails` test helper.

Child processes inherit `ENV`, so you should be able to simply set `ENV["LOGS"]` in the test before calling the `rails` helper.  Also, since the tests are isolated with `ActiveSupport::Testing::Isolation`, you don't need to worry about restoring the original value.

p8 commented on railties/lib/rails/commands/log/log_command.rb:
```diff
@@ -0,0 +1,48 @@
+        end
+
+        def all_environments
+          Dir[Rails::Command.root.join("config/environments/*.rb")]
```
That is a nice change!

JuanVqz commented on railties/test/commands/log_test.rb:
@p8 just wondering if this file name should be `log_command_test.rb` and the class name in line 7 `LogTest` to `LogCommandTest`.

p8 replied:
@JuanVqz Good point.

All existing test files don't have command in the filename, probably because it's already in the directory name.

I think the test classname is currently a bit inconsistent:
https://github.com/rails/rails/blob/38275763257221e381cd4e37958ce1413fd0433c/railties/test/commands/runner_test.rb#L6
https://github.com/rails/rails/blob/38275763257221e381cd4e37958ce1413fd0433c/railties/test/commands/credentials_test.rb#L8

We should probably have rename the classes to the filename.
Do you want to open a PR? :)

JuanVqz replied:
Hey, sure! opened https://github.com/rails/rails/pull/47697, let me know what do you think about it

jonathanhefner commented on railties/CHANGELOG.md:
````diff
@@ -1,3 +1,30 @@
+*   Use Thor for `bin/rails log:clear`.
+
+    With Thor we can't run multiple commands with a single invocation, as
+    additional commands will be seen as arguments to the first command.
+
+    The following was previously possible:
+
+    ```bash
+    bin/rails log:clear tmp:clear
+    ```
+
+    After this change these should be called separately:
+
+    ```bash
+    bin/rails log:clear
+    bin/rails tmp:clear
+    ```
````
Do you think it would be worth adding this to the Upgrading Guide too, for extra visibility?

p8 replied:
I've added a section to the guides.
I've tried to make it a bit more generic.

jonathanhefner commented on railties/lib/rails/commands/log/log_command.rb:
```diff
@@ -0,0 +1,38 @@
+          pattern = environments == ["all"] ? "*" : "{#{environments.join(",")}}"
+          Rails::Command.application_root.glob("log/#{pattern}.log").select(&:file?)
+        end
+
+        def clear_log_file(file)
+          f = File.open(file, "w")
+          f.close
+        end
```
:scissors: No longer using this method.

```suggestion
```

p8 replied:
I've removed the file.

jonathanhefner commented on railties/test/commands/log_test.rb:
```diff
@@ -0,0 +1,93 @@
+
+  test "`rails log:clear` clears all environments log files by default" do
+    Dir.chdir(app_path) do
+      File.open("config/environments/staging.rb", "w")
```
I realize this is just copied from the original test, but this opens a file and leaves the file open until it is collected by GC.  We could use `File.write` like the lines further below, or we could use the `app_file` helper:

```suggestion
      app_file "config/environments/staging.rb", ""
```

If we change all the lines to use the `app_file` and `app_path(_)` helpers, we can drop the `Dir.chdir` block.

p8 replied:
I've refactored the tests to use `app_file`.

jonathanhefner commented on railties/test/commands/log_test.rb:
```diff
@@ -0,0 +1,93 @@
+
+      rails "log:clear"
+
+      assert_equal 0, File.size("log/test.log")
+      assert_equal 0, File.size("log/staging.log")
+      assert_equal 5, File.size("log/dummy.log")
```
I realize this is also copied from the original test, but asserting the exact file size seems overly specific to me.

We could use `assert_not_equal 0`, or we could add private `assert_file_empty` and `assert_file_not_empty` helpers that incorporate `app_path` (in order to drop the `Dir.chdir` block).  e.g.:

```ruby
assert_file_not_empty "log/dummy.log"
```

p8 replied:
I've added assertion for `assert_file_empty` and `assert_file_not_empty`

jonathanhefner commented on railties/test/commands/log_test.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+  end
+
+  test "`rails log:clear test` clears only logs for the specified environment" do
```
For clarity, let's use a placeholder here like:

```suggestion
  test "`rails log:clear some_environment` clears only logs for the specified environment" do
```

p8 replied:
I've changed the test to also test the passing multiple environments.
I hope the following is clear enough:
```ruby
  test "`rails log:clear env1,env2` clears only logs for the specified environments" do
```

jonathanhefner commented on railties/test/commands/log_test.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+  end
+
+  test "`LOGS=test rails log:clear` clears only logs for the specified environment" do
```
Placeholder here too:

```suggestion
  test "`LOGS=some_environment rails log:clear` clears only logs for the specified environment" do
```

p8 replied:
I've changed it to:
```ruby
  test "`LOGS=env1,env2 rails log:clear` clears only logs for the specified environments" do
```

jonathanhefner commented on railties/test/commands/log_test.rb:
```diff
@@ -0,0 +1,93 @@
+
+require "isolation/abstract_unit"
+require "rails/command"
+require "rails/commands/log/log_command"
```
:scissors:

```suggestion
```

p8 replied:
Done.

p8 commented:
@jonathanhefner To keep things backwards compatible I've introduced a `strip_legacy_env_argument` to make the following not break:
```bash
bin/rails log:clear LOGS=all
```

p8 commented:
@jonathanhefner Thanks for the heads up!
I can pick up these changes. No worries.

p8 committed:
Introduce `bin/rails clear` Thor command for clearing logs

Currently we use both Thor and Rake for `bin/rails` commands.
We eventually want to get all the built-in tasks promoted to Thor Commands.
This migrates the `log:clear` task to Thor.

Currently logs and the tmp directory can be cleared with a single invocation:

```bash
bin/rails log:clear tmp:clear
```

With Thor we can't run multiple commands with a single invocation, as
additional commands will be seen as arguments to the first command.

This introduces a single `clear` command that will combine the existing
`log:clear` and `tmp:clear` commands. For now it's limited to clearing logs:

```bash
bin/rails clear --logs
```

The `--logs` option also supports passing `all` or a list of environments:

```bash
bin/rails log:clear --logs all # similar to `bin/rails log:clear LOGS=all`
bin/rails log:clear --logs development test # similar to `bin/rails log:clear LOGS=development,test`
```

Thor doesn't support Rake's FileList, so for truncating all log files we
use a glob pattern and filter out directories.

Co-authored-by: Hartley McGuire <skipkayhil@gmail.com>
Co-authored-by: Jonathan Hefner <jonathan@hefner.pro>

p8 commented:
@jonathanhefner Do you want me to deprecate `rails log:clear` in this PR as well?
I guess it makes more sense once we implement `rails clear --tmp` (or whatever the option will be)?

jonathanhefner commented:
@p8 Yes, I think we can deprecate `log:clear` in this PR.  Then, once this PR is merged, we can implement `clear --tmp` and deprecate `tmp:clear` in the other PR (after a rebase).

jonathanhefner commented on railties/lib/rails/commands/clear/clear_command.rb:
```diff
@@ -0,0 +1,30 @@
+      desc "clear", "Truncate log/*.log files"
+      option :logs, type: :array, lazy_default: [],
+        desc: "Truncate log/*.log files for specified environments to zero bytes. If no environments are specified, truncate logs for all environments."
+      def clear
+        if options[:logs].empty?
+          environments = all_environments
+        else
+          environments = options[:logs]
+        end
+        log_files_for(environments).each { |file| file.truncate(0) }
+      end
+
+      private
+        def log_files_for(environments)
+          pattern = environments == ["all"] ? "*" : "{#{environments.join(",")}}"
+          Rails::Command.application_root.glob("log/#{pattern}.log").select(&:file?)
+        end
+
+        def all_environments
+          Rails::Command.application_root.glob("config/environments/*.rb")
+            .map { |fname| File.basename(fname, ".*") }
+        end
```
Since this is a brand new command, we *can* simplify the "empty" vs "all" distinction.  In other words, let's have `--logs` (with no values specified) truncate all `log/*.log`, and remove the special handling of "all".  (Sorry, I should have called more attention to that in https://github.com/rails/rails/pull/47623#issuecomment-1487364829.)

p8 replied:
@jonathanhefner `all` is slightly different as it will truncate any `log/*.log` instead of `log/{development,test,production}.log`
This was introduced in: https://github.com/rails/rails/issues/22544

jonathanhefner replied:
Yup. :smile:  I was saying that we can remove the old "empty" behavior, and keep the "all" behavior as the new "empty" behavior.

The idea is that:

* It's a bit odd to have `--logs all` behave differently than `--logs foo` or even `--logs ALL`.
* It makes more sense for bare `--logs` to affect every log that could possibly be specified via `--logs ...` (i.e. not limited to environments).
* Targeting a specific subset of logs is still possible via `--logs foo bar etc`.
* If the user doesn't want to clear all logs, it is likely that they want to clear a specific log.  For example, they want to clear `log/test.log` and don't actually care about clearing `log/production.log` (or vice versa).

@rafaelfranca  Do you still agree with keeping only the `LOGS=all` behavior?

p8 replied:
In this case we could remove the whole "checking if an environment even exists" part.
We might have to sanitize the arguments to prevent something like: `--logs ../.env/;`.

For the environments another alternative could be to use the --environment option that is used in other commands:
`bin/rails clear --logs --environment test`
Although that doesn't work with multiple environments.

jonathanhefner replied:
> If we allow passing any log to `--logs`, we could remove the whole "checking if an environment even exists" part.
> We might have to sanitize the arguments to prevent something like: `--logs ../.env/;`.

I might be misunderstanding, but there isn't a check in this PR nor in the Rake task.  And there probably doesn't need to be.  Consider the case where `log/production.log` doesn't exist yet, but a script invokes `bin/rails clear --logs production`.  I don't think that should cause the script to fail.

I also don't think we need to sanitize anything since we're not (or should not be) accepting untrusted input.

> For the environments another alternative could be to use the --environment option that is used in other commands:
> `bin/rails clear --logs --environment test`
> Although that doesn't work with multiple environments (maybe we could do `bin/rails clear --logs --environment local`).

That does not seem preferable to `bin/rails clear --logs test`.  Even `bin/rails clear --logs development test` is shorter than `bin/rails clear --logs --environment local` (39 characters vs 42 characters).  Also, we would still need to support `bin/rails clear --logs foo bar` for arbitrary non-environment log files.

> It would be more consistent with other commands and maybe it would be useful for `clear --tmp` as well.

How would it be useful for `clear --tmp`?  I don't think there are any files in `tmp/` that are segregated by environment, are there?

p8 replied:
> I might be misunderstanding, but there isn't a check in this PR nor in the Rake task. And there probably doesn't need to be. Consider the case where `log/production.log` doesn't exist yet, but a script invokes `bin/rails clear --logs production`. I don't think that should cause the script to fail.

Ah sorry, my mistake. I thought we'd do `log/*.log` or filter input on one of the environments using the following:

https://github.com/rails/rails/blob/6879f89892b4784123fb40cdcc983ac32590a24d/railties/lib/rails/tasks/log.rake#L38-L40

I wasn't aware we can currently do:
```bash
bin/rails log:clear ENV['LOGS']='dummy,puma,redis'
```

> That does not seem preferable to `bin/rails clear --logs test`. Even `bin/rails clear --logs development test` is shorter than `bin/rails clear --logs --environment local` (39 characters vs 42 characters). Also, we would still need to support `bin/rails clear --logs foo bar` for arbitrary non-environment log files.

I think I agree that is preferable.
Also having `--environment` accept multiple environments is inconsistent with other commands.

> How would it be useful for `clear --tmp`? I don't think there are any files in `tmp/` that are segregated by environment, are there?

Hmm, I seem to recall Rails used to cache assets separately in `tmp` but I might be confusing that with the `public/packs` and `public/packs-test` directories.

I do wonder if `assets:clean` should be handled by `clear` as well.
So `clear` would basically clear **all** generated files.
But that is of later concern.
